File: acs-krb-keys-operator/lib/amrc/factoryplus/service_client/discovery.py
# Factory+ Python client library
# Service discovery
# Copyright 2023 AMRC

import logging

from .service_interface     import ServiceInterface
from ..                     import uuids

log = logging.getLogger(__name__)

presets = [
    ("directory_url",       uuids.Service.Directory),
    ("configdb_url",        uuids.Service.ConfigDB),
    ("authn_url",           uuids.Service.Authentication),
    ("mqtt_url",            uuids.Service.MQTT),
];

class Discovery:
    def __init__ (self, fplus):
        self.fplus = fplus

        self.urls = {}
        for key, srv in presets:
            if key in fplus.opts:
                self.set_service_url(srv, fplus.opts[key])
        
    def set_service_url (self, service, url):
        log.info(f"Preset URL for {service}: {url}")
        self.urls[service] = url

    def service_url (self, service):
        url = self.urls.get(service)

        if url is None:
            from_dir = self.fplus.directory.get_service_urls(service)
            if len(from_dir) > 0:
                url = from_dir[0]
                self.urls[service] = url

        return url

File: lib/js-service-client/lib/service/discovery.js
/*
 * Factory+ NodeJS Utilities
 * Service discovery.
 * Copyright 2022 AMRC.
 */

import { Service } from "../uuids.js";

import { ServiceInterface } from "./service-interface.js";

/** Perform service discovery.

The Discovery discovers the URLs used to contact the services. Primarily
this happens via the Directory but facilities are provided to override
URLs for testing.

Options (provided via the ServiceClient) are:

- `directory_url`: (required) The URL of the Directory.
- `authn_url`: The URL of the Auth service.
- `configdb_url`: The URL of the ConfigDB.
- `mqtt_url`: The URL of the MQTT broker.

*/
export class Discovery extends ServiceInterface {
    /** Private. Construct via the ServiceClient. */
    constructor (fplus) {
        super(fplus);
        this.urls = new Map();

        const opts = fplus.opts;
        const presets = [
            [opts.authn_url,        Service.Authentication],
            [opts.configdb_url,     Service.ConfigDB],
            [opts.directory_url,    Service.Directory],
            [opts.mqtt_url,         Service.MQTT],
            [opts.files_url,        Service.Files],
        ];
        for (const [pr, srv] of presets) {
            if (pr == null) continue;
            this.debug.log("discovery", "Preset URL for %s: %s", srv, pr);
            this.set_service_url(srv, pr);
        }
    }

    /** Override a service URL.
     * This is a hook for situations where we must override resolution
     * via the Directory.
     * @arg service A service UUID
     * @arg url The URL to return in future
     */
    set_service_url (service, url) {
        this.urls.set(service, url);
    }

    /** Replace the resolution via the Directory.
     * This is primarily a hook for the Directory itself to use. The
     * function provided must accept a service UUID and return a Promise
     * to an array of URLs. 
     * @arg locater The replacement resolution function
     */
    set_service_discovery (locator) {
        this.find_service_urls = locator;
    }

    /* private */
    find_service_urls (service) {
        return this.fplus.Directory.get_service_urls(service);
    }

    /** Look up the URLs for a service.
     * Although this returns an array, with the current Directory this
     * will never have more than one entry.
     * @arg service A service UUID
     * @returns An Array of URLs.
     * @deprecated
     */
    async service_urls (service) {
        this.debug.log("service", `[${service}] Looking for URL...`);
        if (this.urls.has(service)) {
            const url = this.urls.get(service);
            this.debug.log("service", `[${service}] Found ${url} preconfigured.`);
            return [url];
        }

        const urls = await this.find_service_urls(service);

        if (urls) {
            this.debug.log("service", "[%s] Discovery returned %s",
                service, urls.join(", "));
            return urls;
        } else {
            return [];
        }
    }

    /** Look up the URL for a service.
     * @arg service A service UUID
     * @returns A service URL, or `undefined`.
     */
    async service_url (service) {
        const urls = await this.service_urls(service);
        return urls?.[0];
    }
}
